### PR TITLE
 Input queue

### PR DESCRIPTION
**Pull request by kitadak**

### CODE REVIEW

File: inc/input_queue.hpp [hunk 1]
```diff
@@ -3,38 +3,68 @@
...
 
-
-struct InputEvent 
+struct MouseEvent
 {
-    InputType type;
-    int action = 0;
-    char key = 0;
     double x = 0, y = 0;
+    int button = 0;
```
Comment (eugeneyche) [inline inc/input_queue.hpp:16]: Can we break this into two events?

File: inc/input_queue.hpp [hunk 2]
```diff
@@ -3,38 +3,68 @@
...
 enum class InputType {
     KEYBOARD,
     MOUSE_POS,
     MOUSE_BUTTON,
     MOUSE_SCROLL,
 };
 
-
-struct InputEvent 
+struct MouseEvent
```
Comment (eugeneyche) [inline inc/input_queue.hpp]: Can we do something like `InputMousePressEvent`? It's a little verbose, but I think it'll help with name collisions.

An alternative may be to put it into a separate namespace, but it seems overkill.

File: inc/input_queue.hpp [hunk 3]
```diff
@@ -3,38 +3,68 @@
...
 {
-    InputType type;
-    int action = 0;
-    char key = 0;
     double x = 0, y = 0;
+    int button = 0;
+
+    MouseEvent(double x, double y) :
```
Comment (eugeneyche) [inline inc/input_queue.hpp]:
```
MouseEvent (double x, double y)
  : x (x)
  , y (y)
{ }
```

ditto with all other constructors.

File: inc/input_queue.hpp [hunk 4]
```diff
@@ -3,38 +3,68 @@
...
-    void register_input(InputType, int, char, double, double);
+    InputEvent get_next_event() { InputEvent next = input_queue.front(); input_queue.pop(); return next; }
+    void register_kbd(int key, int scancode, int action, int mods);
+    void register_mouse_button(int button, int action, int mods);
+    void register_mouse_pos(double x, double y);
+    void register_mouse_scroll(double x, double y);
 
 private:
-    std::queue<InputEvent> queue;
+    std::queue<InputEvent> input_queue;
```
Comment (eugeneyche) [inline inc/input_queue.hpp]: private variables should be appended with _ so `input_queue_`

File: inc/input_queue.hpp (hunk 1 above)
Comment (kitadak) [inline inc/input_queue.hpp:16]: Would this be three considering there's mouse press, mouse position, and mouse scroll?

File: src/input_queue.cpp [hunk 5]
```diff
@@ -2,10 +2,28 @@
...
+
+void InputQueue::register_mouse_button(int button, int action, int mods)
+{
+    this->input_queue.push(
+            InputEvent(InputType::MOUSE_BUTTON, MouseEvent(button), action, mods));
+}
+
+void InputQueue::register_mouse_pos(double x = 0, double y = 0)
+{
+    this->input_queue.push(InputEvent(InputType::MOUSE_POS, MouseEvent(x, y)));
```
Comment (eugeneyche) [inline src/input_queue.cpp]: rather than this, can we do just `input_queue_`

**Comment by eugeneyche** [comment]
Awesome! Thanks for the fixes! lgtm, :ship: it!

**Pull request merged**